src/decoding.cc

```diff
@@ -113,11 +115,78 @@ namespace ctranslate2 {
+        StorageView& log_probs) {
+    const dim_t num_beams = logits.dim(0);         
+    const Device device = logits.device();
+    const DataType dtype = logits.dtype();
+
+    if (_spare_beam.dtype() != dtype || _spare_beam.device() != device) {
+      _spare_beam = StorageView(device, dtype);
+    }
+
+    std::vector<StorageView> logit_beam_view_storage(num_beams, StorageView(device, dtype));
+    std::vector<StorageView*> logit_beam_views(num_beams);
+    std::vector<StorageView> log_prob_beam_view_storage(num_beams, StorageView(device, dtype));
+    std::vector<StorageView*> log_prob_beam_views(num_beams);
+    for (int i = 0; i < num_beams; ++i) {
+      logit_beam_views[i] = &(logit_beam_view_storage[i]);
+      log_prob_beam_views[i] = &(log_prob_beam_view_storage[i]);
+    }
+    ops::Split(0, /*no_copy=*/true)(logits, logit_beam_views);
+    log_probs.resize_as(logits);
+    log_probs.reshape(logits.shape());
+    ops::Split(0, /*no_copy=*/true)(log_probs, log_prob_beam_views);
+
+    // Scalar's need to be allocated on CPUs.
+    StorageView scalar_discount(1 - prefix_bias_beta, Device::CPU);
+    // const dim_t cur_beam_size = is_expanded ? _beam_size : 1;
+    assert (num_beams % cur_batch_size == 0);
+    const dim_t cur_beam_size = num_beams / cur_batch_size;
+    for (dim_t b = 0; b < num_beams; ++b) {
+      StorageView &logit_beam = *(logit_beam_views[b]);
+      StorageView &log_prob_beam = *(log_prob_beam_views[b]);
+      const dim_t index_batch = b / cur_beam_size;
+      const dim_t index_beam = b % cur_beam_size;
+      const auto& prefix = prefix_ids[batch_offset[index_batch]];
+      if (static_cast<size_t>(step) < prefix.size() && 
+          !(beams_diverged_from_prefix[index_batch][index_beam])) {
+        ops::SoftMax()(logit_beam, log_prob_beam);
+        ops::Mul()(log_prob_beam, 
+                   scalar_discount.to(log_prob_beam.dtype()),
+                   _spare_beam);
+        const size_t biased_word_id = prefix[step];
+        StorageView spare_scalar_view;
+        TYPE_DISPATCH(
+            _spare_beam.dtype(), 
+            spare_scalar_view = StorageView({1}, _spare_beam.data<T>() + biased_word_id, device));
+        const StorageView spare_scalar_copy(spare_scalar_view);
+        StorageView beta_scalar;
+        TYPE_DISPATCH(
+            _spare_beam.dtype(), 
+            // Scalar's need to be allocated on CPUs.
+            beta_scalar = StorageView(static_cast<T>(prefix_bias_beta), Device::CPU));
+        ops::Add()(spare_scalar_copy, beta_scalar, spare_scalar_view);
+        ops::Log()(_spare_beam, log_prob_beam);
+      } else {
+        ops::LogSoftMax()(logit_beam, log_prob_beam);
+      }
+    }
+  }
 
-  BeamSearch::BeamSearch(const dim_t beam_size, const float length_penalty, const float coverage_penalty)
+  BeamSearch::BeamSearch(const dim_t beam_size, const float length_penalty, const float coverage_penalty, 
+                         const float prefix_bias_beta)
```

Comment:
Minor formatting suggestion:

```suggestion
  BeamSearch::BeamSearch(const dim_t beam_size,
                         const float length_penalty,
                         const float coverage_penalty, 
                         const float prefix_bias_beta)
```